### ISSUE TITLE
X-MimeKit-Warning: Do NOT use ToString() display in console window

### ISSUE THREAD

**Issue by fjames** [jstedfast/MimeKit#623]
The release notes for 2.9.2 (2020-09-12) indicates that "MimeMessage.ToString() now adds an X-MimeKit-Warning header to the beginning of the output string to make it clear to developers doing this that they are Doing it Wrong(tm)."

I am using Serilog for a WebApp project.  When logging to the console the X-MimeKit-Warning appears in the output. I'm using Serilog file based logging as well.  This warning does not appear in the file log.

I believe I'm "Doing it right."

**message.Body = new TextPart("plain")
{
    Text = @"This is a test."
};**

**Here is an example of the console output.
[01:05:54 INF] Mail Queue = 0
[01:06:09 INF] Mail Queue = 1
[01:06:09 INF] Subject: NDM Reports a Connection Warning for (Google.com)
Message:
X-MimeKit-Warning: Do NOT use ToString() to serialize entities! Use one of the WriteTo() methods instead!
Content-Type: text/plain; charset=utf-8

This is a test.**

The Serilog minimum level is set to "Information"

{
  "Serilog": {
    "Using": [ "Serilog.Sinks.Console" ],
    "MinimumLevel": {
      "Default": "Information",
      "Override": {
        "Microsoft": "Warning",
        "System": "Warning"
      }
    },

I need to keep the minimum level set to information so I can log informational items to the console.  I can set the minimum level to warning but that suppresses the informational items.   

Can the "X-MimeKit-Warning" be removed when we are building the body correctly?

This is my first app using MimeKit.  Awesome product so far.

Thanks

**Comment by jstedfast** [comment]
The X-MimeKit-Warning message is not a side-effect of setting the message body, it is a side-effect of calling `message.ToString()`.

Presumably you are doing something like this in your code:

```csharp
Log.Information ("Message: {0}", message);
```

Logging the message like that will invoke ToString() on it which will cause the warning message you are seeing. This warning message will *only* appear in ToString() output and will not be included when you send the message via SmtpClient.

What you'll need to do is find a way to log the message to a stream instead:

```csharp
message.WriteTo (stream);
```

The following docs may provide useful info: https://github.com/serilog/serilog/wiki/Structured-Data. It looks like you might need to register some sort of `Destructure` configuration. I'm not sure how feasible that is, though, since MimeMessage is rather complex.

**Comment by fjames** [comment]
Thanks for the quick reply.  Turns out ToString() is used in some instances even though you don't explicitly add it.
e.g. var t1 = message.Body;    This is actually var t1 = message.Body.ToString();
Thanks for pointing me in the "Doing it right" direction!
--Frank     On Wednesday, November 25, 2020, 06:24:37 AM CST, Jeffrey Stedfast <notifications@github.com> wrote:  
 
 


Closed #623.
